src/diffusers/callbacks.py

```diff
@@ -0,0 +1,116 @@
+from typing import List
+
+from .configuration_utils import ConfigMixin, register_to_config
+from .utils import CONFIG_NAME
+
+
+class PipelineCallback(ConfigMixin):
+    config_name = CONFIG_NAME
+
+    @register_to_config
+    def __init__(self, cutoff_step_ratio=1.0, cutoff_step_index=None):
+        super().__init__()
+
+        if (cutoff_step_ratio is None and cutoff_step_index is None) or (
+            cutoff_step_ratio is not None and cutoff_step_index is not None
+        ):
+            raise ValueError("Either cutoff_step_ratio or cutoff_step_index should be provided, not both or none.")
+
+        if cutoff_step_ratio is not None and (
+            not isinstance(cutoff_step_ratio, float) or not (0.0 <= cutoff_step_ratio <= 1.0)
+        ):
+            raise ValueError("cutoff_step_ratio must be a float between 0.0 and 1.0.")
+
+    @property
+    def tensor_inputs(self) -> List:
+        raise NotImplementedError(f"You need to set the attribute `tensor_inputs` for {self.__class__}")
+
+    def callback_fn(self, pipeline, step_index, timesteps, callback_kwargs):
+        raise NotImplementedError(f"You need to implement the method `callback_fn` for {self.__class__}")
+
+    def __call__(self, pipeline, step_index, timestep, callback_kwargs):
+        return self.callback_fn(pipeline, step_index, timestep, callback_kwargs)
+
+
+class MultiPipelineCallbacks:
```

Comment:
Right now, these callbacks can already be simple functions, right? Can we also wrap them in this class if we want to pass multiple functions? I think it should work, so let's update the type annotation and add a docstring. If it wouldn't work, can we make it work?

Reply:
oh nice idea, I'll take a look into this

Reply:
I'm afraid that this won't be so easy. For the simple callback functions, the users are required to pass the `callback_on_step_end_tensor_inputs` to the pipeline. We can bypass the check I did to overwrite this with the ones from the official callbacks:

```python
  if isinstance(callback_on_step_end, (PipelineCallback, MultiPipelineCallbacks)):
      callback_on_step_end_tensor_inputs = callback_on_step_end.tensor_inputs
```

but the user will need set the inputs correctly since each callback can use different inputs. Without these inputs the callbacks cannot run if we don't change the pipeline code.

IMO I don't think this is worth the extra effort since if the user is going to make a custom callback, he can do everything in just one callback instead of writing multiple ones.

Reply:
Oh too bad, but as you mention not a big loss.